[CDAP-14752] Add search to Metadata SPI

anew opened the pull request:
- adds search to the SPI
- makes the read() method a more generic and extensible
- moves some classes to inner class of MetadataDataset (they are only used there, and they clash with SPI class names)
- removes some unused code
- adds test code for a new bug (to be fixed)

It is best to review the two commits separately.

Commit:
[CDAP-14752] Add search to Metadata SPI, and improve Remove and Read

albertshau commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java:
```diff
@@ -754,7 +751,7 @@ private SearchResults searchByCustomIndex(SearchRequest request) throws BadReque
       @SuppressWarnings("ConstantConditions")
       // stop key should always be the stop key for the namespace, regardless of what the cursor is
-      byte[] stopKey = Bytes.stopKeyForPrefix(namespaceStartKey);
+        byte[] stopKey = Bytes.stopKeyForPrefix(namespaceStartKey);
```
accidental?

anew replied:
funny. IntelliJ indented it that way. Fixed

albertshau commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/MetadataStore.java:
```diff
@@ -51,7 +51,7 @@
    *                             the new metadata contains no or new values for them
    */
-  void replaceMetadata(MetadataScope scope, Metadata metadata,
+  void replaceMetadata(MetadataScope scope, MetadataEntity entity, MetadataDataset.Record metadata,
```
javadoc should be updated.

Also, why does this change need to be made if the record contains the entity? What happens if somebody calls this and the entities don't match?

anew replied:
Good catch. My intention was to just replace this with Record. This must be a relic of an intermediate refactoring I did. Will remove the entity. By the way, I hope to remove this class entirely in the future as it is merely a proxy translating between old and new metadata APIs. The only thing we need to keep is the audit publishing.

rohitsinha54 commented on cdap-data-fabric/src/main/java/co/cask/cdap/spi/metadata/MetadataMutation.java:
```diff
@@ -48,6 +50,14 @@ public MetadataEntity getEntity() {
+   *
+   * Note that the typical use for this is the creation (or recreation) of an
+   * entity, when all the entity's system metadata are defined for the first
+   * time, or redefined.
```
This class seems to used in a Collection which can be set
```
Collection<MetadataChange> batch(Collection<? extends MetadataMutation> mutations) throws IOException;
```

Should provide equals and hashCode implementation

anew replied:
Good point. Added.

rohitsinha54 commented on cdap-data-fabric/src/main/java/co/cask/cdap/spi/metadata/SearchResponse.java:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Since SearchRequest is a member of this class it should override its equals and hashCode too.

anew replied:
good catch. fixed.

rohitsinha54 commented on cdap-data-fabric/src/main/java/co/cask/cdap/spi/metadata/Read.java:
```diff
@@ -0,0 +1,119 @@
+/**
+ * A read from the metadata store.
+ */
+public class Read {
```
Nice 💯

rohitsinha54 commented on cdap-data-fabric/src/main/java/co/cask/cdap/spi/metadata/Sorting.java:
```diff
@@ -0,0 +1,56 @@
+  public Order getOrder() {
+    return order;
+  }
+
```
this should also override equals and hashCode.

anew replied:
done

anew commented:
@albertshau @rohitsinha54 I addressed your comments. Please take another look

rohitsinha54 commented:
LGTM.

albertshau reviewed (approved):
lgtm

anew commented:
Build has previously passed and changes to address comments don't affect tests. Updated copyrights, squashed and rebased on latest develop. Merging now.
